security/pkg/pki/ca/ca.go

```diff
@@ -114,27 +115,31 @@ type IstioCAOptions struct {
 // NewSelfSignedIstioCAOptions returns a new IstioCAOptions instance using self-signed certificate.
 func NewSelfSignedIstioCAOptions(ctx context.Context,
 	rootCertGracePeriodPercentile int, caCertTTL, rootCertCheckInverval, defaultCertTTL,
-	maxCertTTL time.Duration, org string, dualUse bool, namespace string,
-	readCertRetryInterval time.Duration, client corev1.CoreV1Interface,
+	maxCertTTL time.Duration, org string, dualUse bool, namespace string, client corev1.CoreV1Interface,
 	rootCertFile string, enableJitter bool, caRSAKeySize int,
 ) (caOpts *IstioCAOptions, err error) {
+Retry:
 	// For the first time the CA is up, if readSigningCertOnly is unset,
 	// it generates a self-signed key/cert pair and write it to CASecret.
 	// For subsequent restart, CA will reads key/cert from CASecret.
 	caSecret, scrtErr := client.Secrets(namespace).Get(context.TODO(), CASecret, metav1.GetOptions{})
-	if scrtErr != nil && readCertRetryInterval > time.Duration(0) {
-		pkiCaLog.Infof("Citadel in signing key/cert read only mode. Wait until secret %s:%s can be loaded...", namespace, CASecret)
-		ticker := time.NewTicker(readCertRetryInterval)
-		defer ticker.Stop()
-		for scrtErr != nil {
-			select {
-			case <-ticker.C:
-				if caSecret, scrtErr = client.Secrets(namespace).Get(context.TODO(), CASecret, metav1.GetOptions{}); scrtErr == nil {
-					pkiCaLog.Infof("Citadel successfully loaded the secret.")
+	if scrtErr != nil {
+		if !apierror.IsNotFound(scrtErr) {
+			pkiCaLog.Infof("Citadel in signing key/cert read only mode. Wait until secret %s:%s can be loaded...", namespace, CASecret)
+			ticker := time.NewTicker(1 * time.Second)
```

Comment:
This is a big change? As far as I can tell we went from not retrying to retrying. Can you expand on why we need this? What use cases is it fixing?

Reply:
It is to keep the original intention, no strong preference. If we all agree, I can make it fail fast.